File: bots/rs/v55/src/builder/update/turrets.rs
use cambc::EntityType;
use serde_json::{Map, Number, Value};

use crate::builder::Builder;
use crate::util::constants::{INF, MAX_WIDTH};
use crate::util::debug::{Scope, debug as log};
use crate::util::directions::{DIR4, DIR8};

pub fn update_ore_denial(builder: &mut Builder) {
    builder.deny_ore_neighbours = pyrust::set::new!();
    let my_team = builder.state.my_team;

    // Bounded candidate set: union of incrementally-maintained visible
    // ore sets. Typically <10 tiles vs the 69 nearby_tiles we used to
    // walk. The hard cap below ensures pathological many-ore turns
    // don't blow the budget.
    let mut ores: Vec<cambc::Position> = pyrust::vec::new!();
    {
        let _g = Scope::new_timed("ore_collect");
        for p in pyrust::iter!(builder.visible_ti_ores) {
            pyrust::vec::push!(ores, *p);
        }
        for p in pyrust::iter!(builder.visible_ax_ores) {
            pyrust::vec::push!(ores, *p);
        }
    }
    let n_ores = pyrust::len!(ores) as i32;

    let mut n_with_enemy: i32 = 0;
    let mut deep_iters: i32 = 0;
    {
        let _g = Scope::new_timed("ore_scan");
        for pos in &ores {
            let pos = *pos;
            let mut has_enemy = false;
            for d in DIR8 {
                deep_iters += 1;
                let n = pos.add(d);
                if !builder.in_bounds(n) {
                    continue;
                }
                let ni = (n.y as usize) * MAX_WIDTH + (n.x as usize);
                if let Some(team) = builder.building_team[ni]
                    && team != my_team
                {
                    has_enemy = true;
                    break;
                }
                if pyrust::vec::contains!(builder.state.enemy_bots, &n) {
                    has_enemy = true;
                    break;
                }
            }
            if has_enemy {
                n_with_enemy += 1;
                for d in DIR4 {
                    let n = pos.add(d);
                    if builder.in_bounds(n) {
                        pyrust::set::add!(builder.deny_ore_neighbours, n);
                    }
                }
            }
        }
    }

    let n_deny = pyrust::len!(builder.deny_ore_neighbours) as i32;
    let n_enemy_bots = pyrust::len!(builder.state.enemy_bots) as i32;
    let mut args = Map::new();
    pyrust::dict::insert!(
        args,
        pyrust::to_string!("ores"),
        Value::Number(Number::from(n_ores))
    );
    pyrust::dict::insert!(
        args,
        pyrust::to_string!("with_enemy"),
        Value::Number(Number::from(n_with_enemy))
    );
    pyrust::dict::insert!(
        args,
        pyrust::to_string!("deep"),
        Value::Number(Number::from(deep_iters))
    );
    pyrust::dict::insert!(
        args,
        pyrust::to_string!("enemy_bots"),
        Value::Number(Number::from(n_enemy_bots))
    );
    pyrust::dict::insert!(
        args,
        pyrust::to_string!("deny"),
        Value::Number(Number::from(n_deny))
    );
    log(
        "ore_deny: ores={ores} with_enemy={with_enemy} deep={deep} enemy_bots={enemy_bots} deny={deny}",
        args,
    );
}

pub fn update_enemy_turrets(builder: &mut Builder) {
    let my_team = builder.state.my_team;
    if let Some(t) = builder.nearest_enemy_turret {
        let i = (t.y as usize) * MAX_WIDTH + (t.x as usize);
        let valid = matches!(
            builder.building_kind[i],
            Some(EntityType::Gunner | EntityType::Sentinel)
        ) && builder.building_team[i] != Some(my_team);
        if !valid {
            builder.nearest_enemy_turret = None;
        }
    }

    let mut min_dist = INF;
    let nearby = pyrust::clone!(builder.state.nearby_tiles);
    for pos in &nearby {
        let i = (pos.y as usize) * MAX_WIDTH + (pos.x as usize);
        let is_enemy_turret = matches!(
            builder.building_kind[i],
            Some(EntityType::Gunner | EntityType::Sentinel)
        ) && builder.building_team[i] != Some(my_team)
            && pyrust::is_some!(builder.building_team[i]);
        if is_enemy_turret {
            let dist = builder.state.my_pos.distance_squared(*pos);
            if dist < min_dist {
                min_dist = dist;
                builder.nearest_enemy_turret = Some(*pos);
            }
        }
    }
}

File: bots/beacon/bestrn/builder/update/turrets.py
from __future__ import annotations

from cambc import EntityType, Environment
from typing import TYPE_CHECKING
if TYPE_CHECKING:
    from builder import Builder
from util.constants import INF, MAX_WIDTH
from util.directions import DIR4, DIR8

def update_ore_denial(builder):
    builder.deny_ore_neighbours = set()
    nearby = list(builder.state.nearby_tiles)
    my_team = builder.state.my_team
    for pos in nearby:
        env = builder.env[int(pos.y) * 50 + int(pos.x)]
        if env != Environment.ORE_TITANIUM and env != Environment.ORE_AXIONITE:
            continue
        has_enemy = False
        for d in DIR8:
            n = pos.add(d)
            if not builder.in_bounds(n):
                continue
            ni = int(n.y) * 50 + int(n.x)
            team = builder.building_team[ni]
            if team is not None and (team != my_team):
                has_enemy = True
                break
            if (n in builder.state.enemy_bots):
                has_enemy = True
                break
        if has_enemy:
            for d in DIR4:
                n = pos.add(d)
                if builder.in_bounds(n):
                    builder.deny_ore_neighbours.add(n)

def update_enemy_turrets(builder):
    my_team = builder.state.my_team
    t = builder.nearest_enemy_turret
    if t is not None:
        i = int(t.y) * 50 + int(t.x)
        valid = ((builder.building_kind[i] is not None) and (builder.building_kind[i] == EntityType.GUNNER or builder.building_kind[i] == EntityType.SENTINEL)) and builder.building_team[i] != my_team
        if not valid:
            builder.nearest_enemy_turret = None
    min_dist = 1000000
    nearby = list(builder.state.nearby_tiles)
    for pos in nearby:
        i = int(pos.y) * 50 + int(pos.x)
        is_enemy_turret = ((builder.building_kind[i] is not None) and (builder.building_kind[i] == EntityType.GUNNER or builder.building_kind[i] == EntityType.SENTINEL)) and builder.building_team[i] != my_team and (builder.building_team[i] is not None)
        if is_enemy_turret:
            dist = builder.state.my_pos.distance_squared(pos)
            if dist < min_dist:
                min_dist = dist
                builder.nearest_enemy_turret = pos
